theia-ide/theia
# complete support of variable substitution

#### What it does

- fix #3434 - complete support of vscode variable substitution
  - only [input variables for tasks](https://code.visualstudio.com/docs/editor/variables-reference#_input-variables) are left out, since it depends on #5013 and requires substantial refactoring of the task extension first
- fixed resolution of env variables, see https://github.com/theia-ide/theia/pull/5835/commits/fa3016e44c0b143530aa763dd217437c5f1f7e58
- [x] a CQ for copied input schema and types: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20504

#### How to test

- look at https://code.visualstudio.com/docs/editor/variables-reference and test different variables
- you can also follow commit history to see support for which variables were added
- for most variables, you can use `Variables: List All` command
  - then select a variable to test
  - then you will be prompted to provide an argument for a variable, an argument is a part of ':' in variable substitution expression
  - if a variable does not accept argument just press enter, if it accepts then provide a proper argument
  - you should see a variable value if it can be computed in the given context, or raw expression if it cannot
- alternatively, you can define variables in tasks.json and launch.json and then try to run a task or launch configuration
  - like defined [here](https://code.visualstudio.com/docs/editor/variables-reference#_why-isnt-workspaceroot-documented) for example, there is a bug though that an output for short-lived task is lost, you may want to add `watch echo` in order to keep a task running
- for [command variables](https://code.visualstudio.com/docs/editor/variables-reference#_command-variables) you may want to install vscode node js extensions instead of native and test with `extension.pickNodeProcess`, don't forget that you need both vscode nodejs extension to use them
- for input variables you can follow: https://code.visualstudio.com/docs/editor/variables-reference#_input-variables
  - tasks.json is not supported, since the task extension is not ready for it, see #5836

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#reviewing)
